Allow unblocking of non-subscribed signals

ftilde opened the pull request:
This should fix #2.

Calling the newly introduced function `unblock_signals_by_default` resets the mask of the process to an empty one. This has to be done before calling `notify` or `notify_on` (which is enforced at runtime). Subscribing to a signal via `notify_on` now (again) blocks the current signal, allowing it to be picked by `sigwait` in the watcher thread.

This works because if a signal arrives asynchronously from another process, it will be passed to one (arbitrary) thread that has not masked it. Only if there is no such thread (e.g., if one has subscribed to that signal via `notify_on`), it will be enqueued as pending and can then be picked up by `sigwait`. It should be noted that the worker thread still blocks and waits for all subscribable signals, but will just never receive a signal that is not also blocked by all other threads.

A nicer API would certainly be to not block any signals by default and maybe allow for opt-in blocking of certain signals via separate functions, but that would be a breaking change.

ftilde committed:
Allow unblocking of non-subscribed signals

- It is possible to block only those signals that are passed as parameters to
  notify/notify_on in the main thread and all threads spawned later.
- In order not to change the default behavior, all subscribable signals
  are still blocked by default. The new behavior can be opted into by
  calling the new function unblock_signals_by_default before
  notify/notify_on.

BurntSushi commented:
> A nicer API would certainly be to not block any signals by default and maybe allow for opt-in blocking of certain signals via separate functions, but that would be a breaking change.

@ftilde Breaking changes are OK. We can just release a semver bump.

BurntSushi commented:
@ftilde What do you think about coming up with a better API and then we can release `0.3.0` instead of merging this?

I think I would also be OK merging this and releasing a new patch release if you just want a quick fix.

ftilde committed:
Fix blocking of multiple signals per pipe

- Also update the example test application to catch a similar error on
  execution.

ftilde commented:
I just fixed a stupid error. (Somehow I thought the map had signals as keys.)

@BurntSushi If you don't mind a semver bump, I can go ahead and change the default to not block any signals and add additional `block_*` functions.

ftilde committed:
Do not block any signals by default

- Signals are added to the mask of blocked signals lazily once they are
  subscribed to.
- Users can block signals manually (without subscribing to a signal via
  a channel) by the block function
- The previous behavior is restored by calling block_all_subscribable
  at the beginning of the program.

ftilde commented:
@BurntSushi Ok, in the latest commit I now have changed the default behavior to not block any signals. However, users can mask a list of signals explicitly using `block`. Also, the previous default blocking behavior can be conveniently restored by calling `block_all_subscribable` at the beginning of the program.

Feel free to merge and bump the version to `0.3.0` if you are OK with the changes.

BurntSushi commented:
All right, this looks great, thank you!

BurntSushi commented:
`chan-signal 0.3.0` is on crates.io!

ftilde commented:
Excellent!
